src/source-map.ts

```diff
@@ -0,0 +1,216 @@
+ * subject to an additional IP rights grant found at
+ * http://polymer.github.io/PATENTS.txt
+ */
+
+import * as acorn from 'acorn';
+import * as dom5 from 'dom5';
+import * as parse5 from 'parse5';
+import {Analyzer} from 'polymer-analyzer';
+import {AnalysisContext} from 'polymer-analyzer/lib/core/analysis-context';
+import {ParsedHtmlDocument} from 'polymer-analyzer/lib/html/html-document';
+import {RawSourceMap, SourceMapConsumer, SourceMapGenerator} from 'source-map';
+import * as urlLib from 'url';
+
+import * as matchers from './matchers';
+
+const inlineSourcemapPrefix =
+    '\n//# sourceMappingURL=data:application/json;charset=utf8;base64,';
+
+const sourceMappingUrlExpr = /\n\/\/# sourceMappingURL=(.*)\n?/;
+const inlineSourceMapExpr =
+    /^data:application\/json;(charset=[^;]+;)?base64,([a-zA-Z0-9+\/=]+)$/;
+
+
+function base64StringToRawSourceMap(input: string) {
+  return JSON.parse(
+      Buffer.from(input, 'base64').toString('utf8')) as RawSourceMap;
+}
+
+function rawSourceMapToBase64String(sourcemap: RawSourceMap) {
+  return Buffer.from(JSON.stringify(sourcemap), 'utf8').toString('base64');
+}
+
+/**
+ * Creates an identity source map from JS script content. Can offset original
+ * line/column data for inline script elements.
+ */
+function createJsIdentitySourcemap(
+    sourceUrl: string,
+    sourceContent: string,
+    lineOffset: number,
+    firstLineCharOffset: number) {
+  const generator = new SourceMapGenerator();
+  const tokenizer = acorn.tokenizer(sourceContent, {locations: true});
+  for (let token = tokenizer.getToken(); token.type.label !== 'eof';
+       token = tokenizer.getToken()) {
+    if (!token.loc) {
+      return null;
+    }
+    let mapping: any = {
+      original: {
+        line: token.loc.start.line + lineOffset,
+        column: token.loc.start.column +
+            (token.loc.start.line === 1 ? firstLineCharOffset : 0)
+      },
+      generated: token.loc.start,
+      source: sourceUrl
+    };
+
+    if (token.type.label === 'name') {
+      mapping.name = token.value;
+    }
+
+    generator.addMapping(mapping);
+  }
+
+  return generator.toJSON();
+}
+
+function offsetSourceMap(
+    sourcemap: RawSourceMap, lineOffset: number, firstLineCharOffset: number) {
+  const consumer = new SourceMapConsumer(sourcemap);
+  const generator = new SourceMapGenerator();
+
+  consumer.eachMapping(mapping => {
+    const newMapping: any = {
+      source: mapping.source,
+      original: {line: mapping.originalLine, column: mapping.originalColumn},
+      generated: {
+        line: mapping.generatedLine + lineOffset,
+        column: mapping.generatedColumn +
+            (mapping.generatedLine === 1 ? firstLineCharOffset : 0)
+      }
+    };
+
+    if (mapping.name) {
+      newMapping.name = mapping.name;
+    }
+
+    generator.addMapping(newMapping);
+  });
+
+  return generator.toJSON();
+}
+
+export async function getExistingSourcemap(
+    analyzer: AnalysisContext|Analyzer,
```

Comment:
`AnalyzerContext` was coming from passing `document.analyzer` to this method. I suppose I could typecast that call ...